### PR TITLE
Fixes #3 - ArgumentOutOfRangeException in ConstantMutator

### LINKED ISSUE #3: intermittent "index out of range" on preview
(keepass v2.45, this plugin v1.2.0)

I see the following error in the preview window, but only intermittently - I either see a list of passwords or this error.

```
Failed to generate a password.
 
Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
```

If I use the plugin as the default password generator, this just results in intermittently not getting a password at all in the password field.

The previous version I used (I think it was 1.1.2 but I'm not certain) would, with some more restrictive settings, give a list with some entries and others with just an error text along the lines of "could not generate" (sorry again, I don't remember the exact text). With this version it's more all or nothing.

If the new approach is by design, I think the previous approach was preferable.

### PR DESCRIPTION
**Pull request by AGBrown**
This should fix #3 - the `ArgumentOutOfRangeException` in `ConstantMutator`

I've created a unit test project and added a test fixture for the `ConstantMutator` and submitted a fix for the same.

I've had to make some guesses about how `ConstantMutator` is meant to behave. I've assumed

1. The input `StringBuilder` may or may not have a trailing space depending on what mutators have done before this one is invoked, e.g. `"foo"` → `". foo"`
2. If the mutation is to be made at the start, or middle of the phrase then the mutation is inserted with a trailing whitespace separator, e.g. `"foo bar fizz bang"` → `"foo . bar fizz bang"`.
3. If the mutation is to be made at the end of the phrase then the mutation is inserted without a trailing whitespace separator, and before any whitespace that exists, e.g. `"foo 6"` → `"foo 6."` and `"foo "` → `"foo. "`

Assumption (3) in particular seems like it might be wrong, so needs reviewing, but is easy enough to change the tests and code if it needs correcting.

The penultimate commit (`4e8f38b`) will fail the relevant test, and the final commit (`bf246f4`) makes the tests pass by fixing the `Mutate` method. The rest of the commits are just setting up the test infrastructure as there wasn't a unit test project already in place (unless I missed it).

### CODE REVIEW

**Comment by AGBrown** [comment]
I just had an intermittent test failure on `Mutate_ForAnywhere_AddsMutation_BetweenWords`. I'm not sure why, so I'll do test runs until failure again to see if it's an issue with the test fixtures or something else.

**Comment by AGBrown** [comment]
It's an issue with the encoding of the test words file `scowl wordlist up to 50.txt` in the RandomiseWordsList project. I'll look at a fix for it later/tomorrow.

**Comment by ligos** [comment]
Thanks @AGBrown - I'm starting work now, but I'll check your changes in more detail tonight.

**Comment by AGBrown** [comment]
Encoding issue fixed with `140cf3b`. The file was encoded as windows 1252 after my git clone (according to sublime, which seemed to read it all correctly as compared to the tests which were reading many characters in as ). As the code I grabbed from `RandomiseWordsList` used `File.OpenText` it was using UTF8 encoding, therefore the accented characters were read as `U+FFFC` which failed `Char.IsLetterOrDigit` and so the word boundaries were incorrect compared to what was expected by the tests.

I added the nugets to the test project, updated the `StreamReader` to use the right encoding and now it reads the accented characters correctly. A 10 minute "run unit tests until fail" run didn't fail so I'm as sure as I can be that this must have been the cause of the error.

TL;DR: the failure didn't really related to the test itself but this was the simplest way to fix and move on for the moment.

**Comment by ligos** [comment]
Thanks for your work, Andrew. I won't merge this pull request (as appealing as having more than zero unit tests is, I doubt I'll have time to improve coverage). I've responded to the original issue.

**Comment by ligos** [comment]
PS: you were right about assumption 3 - the `EndOfWord` is a special case to make it look more "natural" without whitespace preceding the `.`. So `I am a sentence.` rather than `I am a sentence .`

**Comment by ligos** [comment]
Closing as fix was applied via other means.
